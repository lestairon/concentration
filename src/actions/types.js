export const CREATE_BOARD = "CREATE_BOARD";
export const FLIP_CARD = "FLIP_CARD";
export const TOGGLE_DISABLED = "TOGGLE_DISABLED";
export const SET_SOLVED = "SET_SOLVED";
export const INCREMENT_MOVES = "INCREMENT_MOVES";
export const RESET_FLIPPED = "RESET_FLIPPED";
export const TOGGLE_MENU = "TOGGLE_MENU";
