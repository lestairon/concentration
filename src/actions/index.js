import * as actions from "../constants/actionTypes";

const createBoard = ({ pairOfCards, ordered }) => ({
  type: actions.CREATE_BOARD,
  pairOfCards,
  ordered
});

const flipCard = id => ({
  type: actions.FLIP_CARD,
  id
});

const toggleDisabled = () => ({
  type: actions.TOGGLE_DISABLED
});

const setSolved = cards => ({
  type: actions.SET_SOLVED,
  cards
});

const incrementMoves = () => ({
  type: actions.INCREMENT_MOVES
});

const resetFlipped = () => ({
  type: actions.RESET_FLIPPED
});

const toggleMenu = () => ({
  type: actions.TOGGLE_MENU
});

const runTimer = () => {
  return (dispatch, getState) => {
    dispatch({ type: actions.RUN_TIMER, time: Date.now() });
    let timer = () => {
      if (getState().timer.running) {
        dispatch({ type: actions.TIME, time: Date.now() });
        setTimeout(timer, 1000);
      }
    };
    timer();
  };
};

const stopTimer = () => ({
  type: actions.STOP_TIMER
});

export {
  createBoard,
  toggleDisabled,
  setSolved,
  incrementMoves,
  toggleMenu,
  flipCard,
  resetFlipped,
  runTimer,
  stopTimer
};
