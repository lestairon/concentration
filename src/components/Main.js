import React from "react";
import MoveCounter from "./MoveCounter";
import Stopwatch from "./Stopwatch";
import PropTypes from "prop-types";
import { useSelector } from "react-redux";
import { StyledMain } from "./components";
import Game from "./Game";

const Main = () => {
  const { moveCount } = useSelector(({ boardState }) => boardState);

  return (
    <StyledMain>
      <MoveCounter moveCount={moveCount} />
      <Game />
      <Stopwatch />
    </StyledMain>
  );
};

Main.propTypes = {
  order: PropTypes.bool
};

export default Main;
