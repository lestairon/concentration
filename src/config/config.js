import firebase from "firebase/app";
import "firebase/firestore";

const firebaseConfig = {
  apiKey: "AIzaSyAqp6h3M478xGTxf_fa5sqQHcuMjS0hqOc",
  authDomain: "concentration-64b0c.firebaseapp.com",
  databaseURL: "https://concentration-64b0c.firebaseio.com",
  projectId: "concentration-64b0c",
  storageBucket: "",
  messagingSenderId: "487946285632",
  appId: "1:487946285632:web:99ec7eef3a53e26a"
};

firebase.initializeApp(firebaseConfig);

export default firebase;
