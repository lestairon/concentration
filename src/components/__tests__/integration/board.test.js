import * as actions from "../../../actions";
import Main from "../../Main";
import React from "react";
import { render, fireEvent, waitForDomChange } from "@testing-library/react";
import { Provider } from "react-redux";
import { createStore, applyMiddleware, compose } from "redux";
import rootReducer from "../../../reducers";
import thunk from "redux-thunk";
import "jest-styled-components";
import { getFirestore, reduxFirestore } from "redux-firestore";
import config, { firestoreDB } from "../../../config/config";

const component = (
  ui,
  {
    initialState,
    store = createStore(
      rootReducer,
      initialState,
      compose(
        applyMiddleware(thunk.withExtraArgument({ getFirestore })),
        reduxFirestore(config)
      )
    )
  } = {}
) => {
  return { ...render(<Provider store={store}>{ui}</Provider>), store };
};

describe("board responds correctly to state", () => {
  it("finishes the game correctly", async () => {
    const { container, getByText } = component(<Main numberOfCards={1} />);
    const cards = container.firstChild.firstChild.nextSibling;
    fireEvent.click(cards.firstChild);
    fireEvent.click(cards.lastChild);
    await waitForDomChange({ container });
    expect(getByText(/restart/i)).toBeTruthy();
  });

  it("restarts the game", async () => {
    const { container, getByText } = component(<Main numberOfCards={1} />);
    const cards = container.firstChild.firstChild.nextSibling;
    fireEvent.click(cards.firstChild);
    fireEvent.click(cards.lastChild);
    await waitForDomChange({ container });
    fireEvent.click(getByText(/restart/i));
    expect(container).toMatchSnapshot();
  });

  it("updates the move counter", () => {
    const { container } = component(<Main numberOfCards={1} />);
    expect(container.firstChild.firstChild.innerHTML).toBe(
      "Number of moves: 0"
    );
    const cards = container.firstChild.firstChild.nextSibling;
    fireEvent.click(cards.firstChild);
    fireEvent.click(cards.lastChild);
    expect(container.firstChild.firstChild.innerHTML).toBe(
      "Number of moves: 1"
    );
  });

  it("disables the board when 2 cards are flipped", () => {
    const { container } = component(<Main numberOfCards={2} />);
    const cards = container.firstChild.firstChild.nextSibling;
    fireEvent.click(cards.firstChild);
    fireEvent.click(cards.lastChild);
    fireEvent.click(cards.firstChild.nextSibling);
    expect(cards.firstChild.nextSibling.textContent).toBe("");
  });

  it("shows the card for a short span of time if incorrect", async () => {
    jest.useFakeTimers();
    const { container } = component(<Main numberOfCards={5} ordered={true} />);
    const cards = container.firstChild.firstChild.nextSibling;
    jest.spyOn(actions, "runTimer").mockImplementation(() => ({ type: null }));
    fireEvent.click(cards.firstChild);
    fireEvent.click(cards.lastChild);
    expect(setTimeout).toHaveBeenCalledWith(expect.any(Function), 800);
    const spyResetFlipped = jest.spyOn(actions, "resetFlipped");
    jest.runAllTimers();
    expect(spyResetFlipped).toHaveBeenCalled();
    jest.useRealTimers();
  });
});

describe("score board functionality", () => {
  const mock = score =>
    jest.spyOn(firestoreDB, "collection").mockImplementation(() => ({
      limit: () => ({
        onSnapshot: cb => {
          setTimeout(() =>
            cb({
              docChanges: () => [
                {
                  type: "added",
                  doc: { data: () => ({ name: "test", score }) }
                }
              ]
            })
          );
        }
      }),
      add: () => Promise.resolve()
    }));

  it("fetch the score ", async () => {
    mock(1);
    const { container } = component(<Main numberOfCards={0} />);
    await waitForDomChange({ container });
    const scoreBoard = container.firstChild.firstChild.nextSibling.firstChild;
    expect(scoreBoard).toMatchSnapshot();
  });

  it("updates after submitting new score", async () => {
    mock(420);
    const { container, queryByText } = component(<Main numberOfCards={1} />);
    const cards = container.firstChild.firstChild.nextSibling;
    fireEvent.click(cards.firstChild);
    fireEvent.click(cards.lastChild);
    await waitForDomChange({ container });
    const submitButton = queryByText(/submit score/i);
    expect(submitButton).toBeTruthy();
    fireEvent.click(submitButton);
    await waitForDomChange({ container });
    const scoreBoard = container.firstChild.firstChild.nextSibling.firstChild;
    expect(scoreBoard).toMatchSnapshot();
  });
});
