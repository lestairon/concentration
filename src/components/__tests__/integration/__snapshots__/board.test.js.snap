// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`board responds correctly to state restarts the game 1`] = `
.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: space-around;
  -webkit-justify-content: space-around;
  -ms-flex-pack: space-around;
  justify-content: space-around;
  height: 80vh;
  width: 20%;
  margin: auto;
}

.c0 > label {
  position: absolute;
  top: calc(34% - 1em);
  font-family: Nunito;
}

.c0 > button {
  box-shadow: 0 0.5rem 0.5rem rgba(0,0,0,0.2);
  padding: 10px 10px;
  -webkit-letter-spacing: 1px;
  -moz-letter-spacing: 1px;
  -ms-letter-spacing: 1px;
  letter-spacing: 1px;
  font-family: Quicksand;
  font-weight: bold;
  border-radius: 4px;
  outline: none;
  cursor: pointer;
  -webkit-transition: all linear 0.2s;
  transition: all linear 0.2s;
  color: black;
}

.c0 > button:hover {
  background-color: mediumaquamarine;
  -webkit-transform: scale(1.02);
  -ms-transform: scale(1.02);
  transform: scale(1.02);
}

<div>
  <form
    class="c0"
  >
    <label>
      Select a difficulty
    </label>
    <button
      data-testid="button"
      style="display: none;"
      value="test"
    />
    <button
      class="easy"
      value="easy"
    >
      Easy
    </button>
    <button
      class="advanced"
      value="advanced"
    >
      Advanced
    </button>
    <button
      class="hard"
      value="hard"
    >
      Hard
    </button>
  </form>
</div>
`;

exports[`score board functionality fetch the score  1`] = `
.c0 {
  font-family: Nunito,sans-serif;
  grid-column-start: 1;
  grid-column-end: 3;
  grid-row-start: 1;
  grid-row-end: 2;
  list-style: none;
  text-align: center;
  padding-left: 0;
}

.c1::before {
  content: "";
  width: 65%;
  display: block;
  margin: auto;
  border-top: 1px solid rgba(0,0,0,0.3);
}

<ol
  class="c0"
>
  Top 3 players:
  <li
    class="c1"
  >
    Player: test
            Score: 1
  </li>
</ol>
`;

exports[`score board functionality updates after submitting new score 1`] = `
.c0 {
  font-family: Nunito,sans-serif;
  grid-column-start: 1;
  grid-column-end: 3;
  grid-row-start: 1;
  grid-row-end: 2;
  list-style: none;
  text-align: center;
  padding-left: 0;
}

.c1::before {
  content: "";
  width: 65%;
  display: block;
  margin: auto;
  border-top: 1px solid rgba(0,0,0,0.3);
}

<ol
  class="c0"
>
  Top 3 players:
  <li
    class="c1"
  >
    Player: test
            Score: 420
  </li>
</ol>
`;
